src/modules/rlm_ldap/rlm_ldap.c

```diff
@@ -2035,6 +2177,73 @@ static int mod_instantiate(module_inst_ctx_t const *mctx)
 	return -1;
 }
 
+static const module_env_t sasl_module_env[] = {
+	{ FR_MODULE_ENV_OFFSET("mech", FR_TYPE_STRING, ldap_auth_mod_env_t, user_sasl_mech,
+			       NULL, T_INVALID, false, false, false) },
+	{ FR_MODULE_ENV_OFFSET("authname", FR_TYPE_STRING, ldap_auth_mod_env_t, user_sasl_authname,
+			       NULL, T_INVALID, false, false, false) },
+	{ FR_MODULE_ENV_OFFSET("proxy", FR_TYPE_STRING, ldap_auth_mod_env_t, user_sasl_proxy,
+			       NULL, T_INVALID, false, true, false) },
+	{ FR_MODULE_ENV_OFFSET("realm", FR_TYPE_STRING, ldap_auth_mod_env_t, user_sasl_realm,
+			       NULL, T_INVALID, false, true, false) },
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_env_t auth_user_module_env[] = {
+	{ FR_MODULE_ENV_OFFSET("base_dn", FR_TYPE_STRING, ldap_auth_mod_env_t, user_base,
+			       "", T_SINGLE_QUOTED_STRING, true, false, true) },
+	{ FR_MODULE_ENV_OFFSET("filter", FR_TYPE_STRING, ldap_auth_mod_env_t, user_filter,
+			       NULL, T_INVALID, false, true, true) },
+	{ FR_MODULE_ENV_SUBSECTION("sasl", NULL, sasl_module_env) },
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_env_t authenticate_module_env[] = {
+	{ FR_MODULE_ENV_SUBSECTION("user", NULL, auth_user_module_env) },
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_env_t autz_user_module_env[] = {
+	{ FR_MODULE_ENV_OFFSET("base_dn", FR_TYPE_STRING, ldap_autz_mod_env_t, user_base,
+			       "", T_SINGLE_QUOTED_STRING, true, false, true) },
+	{ FR_MODULE_ENV_OFFSET("filter", FR_TYPE_STRING, ldap_autz_mod_env_t, user_filter,
+			       NULL, T_INVALID, false, true, true) },
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_env_t autz_group_module_env[] = {
+	{ FR_MODULE_ENV_OFFSET("base_dn", FR_TYPE_STRING, ldap_autz_mod_env_t, group_base,
+			       NULL, T_INVALID, false, false, true) },
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_env_t autz_profile_module_env[] = {
+	{ FR_MODULE_ENV_OFFSET("default", FR_TYPE_STRING, ldap_autz_mod_env_t, default_profile,
+			       NULL, T_INVALID, false, false, true) },
+	{ FR_MODULE_ENV_OFFSET("filter", FR_TYPE_STRING, ldap_autz_mod_env_t, profile_filter,
+			       "(&)", T_SINGLE_QUOTED_STRING, false, false, true ) },	//!< Correct filter for when the DN is known.
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_env_t authorize_module_env[] = {
+	{ FR_MODULE_ENV_SUBSECTION("user", NULL, autz_user_module_env) },
+	{ FR_MODULE_ENV_SUBSECTION("group", NULL, autz_group_module_env) },
+	{ FR_MODULE_ENV_SUBSECTION("profile", NULL, autz_profile_module_env) },
+	MODULE_ENV_TERMINATOR
+};
+
+static const module_method_env_t authenticate_method_env = {
+	.inst_size = sizeof(ldap_auth_mod_env_t),
+	.inst_type = "ldap_auth_mod_env_t",
+	.env = authenticate_module_env
+};
+
+static const module_method_env_t authorize_method_env = {
+	.inst_size = sizeof(ldap_autz_mod_env_t),
+	.inst_type = "ldap_autz_mod_env_t",
+	.env = authorize_module_env
+};
```

Comment:
Can these live at the top of the file with the CONF_PARSER sections, preferably interleaved with them...